### PR TITLE
l4zBlwLF Implemented caching on the remote config.

### PR DESCRIPTION
**Pull request by j-maynard**
Implemented caching on the remote config and setup the getConfig
method to throw IOEXceptions which can be caught to fallback to
the normal fed-config.

### CODE REVIEW

File: hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java [hunk 1]
```diff
@@ -17,15 +18,23 @@
 import java.io.InputStream;
 import java.text.DateFormat;
 import java.text.SimpleDateFormat;
+import java.util.Date;
 
 public class S3ConfigSource {
     private static final Logger LOG = LoggerFactory.getLogger(S3ConfigSource.class);
     private ConfigConfiguration configConfiguration;
-    private AmazonS3 s3Client;
+    private S3Object s3Object;
```
Comment (kerrr) [inline hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java]: This doesn't look right to me. Does this work? Shouldn't we be getting the s3Object every request.

File: hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java [hunk 2]
```diff
@@ -17,15 +18,23 @@
...
+    private long cacheLengthInMillis;
+    private Date lastModifiedDate;
+    private Long lastAccessedTime;
+    private RemoteConfigCollection currentConfig;
 
     public S3ConfigSource(ConfigConfiguration configConfiguration, AmazonS3 s3Client) {
         this.configConfiguration = configConfiguration;
-        this.s3Client = s3Client;
+        this.cacheLengthInMillis = configConfiguration.getSelfService().getCacheLengthInSeconds() * 1000;
+        this.s3Object = s3Client.getObject(
```
Comment (kerrr) [inline hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java]: see above comment. This feels wrong.

File: hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java [hunk 3]
```diff
@@ -49,33 +58,61 @@ public static S3ConfigSource setupS3ConfigSource(ConfigConfiguration configConfi
...
         } finally {
-            if(fullObject != null) {
+            if(this.s3Object != null) {
                 try {
-                    fullObject.close();
+                    this.s3Object.close();
                 } catch(IOException e) {
                     throw new RuntimeException(e);
                 }
             }
```
Comment (kerrr) [inline hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java]: I think we could do a better job of the exception handling here. If we're going to throw the IOException anyway, they we shouldn't need all this.

**Comment by chrisholmes** [comment]
We've got existing prior art for using a [`LoadingCache`](https://github.com/google/guava/wiki/CachesExplained#from-a-cacheloader) for this kind of stuff. It should make this a lot simpler and we won't have to think about concurrency isssue so much.

**Comment by jakubmiarka** [comment]
Should this be closed in favour of https://github.com/alphagov/verify-hub/pull/300 ?

**Comment by kerrr** [comment]
Yes. It's been superseded.

**Pull request closed without merging**